src/NServiceBus.Core/Persistence/PersistenceExtentions.cs

```diff
@@ -85,14 +86,31 @@ public PersistenceExtentions(Type definitionType, SettingsHolder settings, Type
                 SelectedStorages = new List<Type>(),
             };
 
+            
             if (storageType != null)
             {
+                var storageSupported = StorageTypeSupported(definitionType, storageType);
+                if (!storageSupported)
+                {
+                    throw new Exception(string.Format("Persistence '{0}' does not support storage type {1}.", definitionType.Name, storageType.Name));
+                }
+
                 enabledPersistence.SelectedStorages.Add(storageType);
             }
 
             definitions.Add(enabledPersistence);
         }
 
+        private static bool StorageTypeSupported(Type definitionType, Type storageType)
```

Comment:
I think we should rather add a non-generic variant of HasSupportFor. The generic param there seems to be only a syntactic sugar that in this case makes things worse. This way we won't need reflection.